internal/updater/update.go

```diff
@@ -1,285 +1,101 @@
+		return err
 	}
-	return nil
-}
 
-func updateTo(ctx context.Context, version, url string) error {
-	debug.Log("URL: %s", url)
-	out.Green(ctx, "Update available!")
-	if true {
-		out.Red(ctx, "Gopass Updater is disabled (insecure, see #1676")
-		return nil
-	}
-	ok, err := termio.AskForBool(ctx, fmt.Sprintf("Do you want to update gopass to %s?", version), true)
+	debug.Log("verifying GPG signature ...")
+	ok, err := gpgVerify(sha256sums, sig)
 	if err != nil {
-		return err
+		return errors.Wrapf(err, "signature verification failed")
 	}
 	if !ok {
-		return nil
+		return fmt.Errorf("GPG signature verification for SHA256SUMS failed")
 	}
-	return updateGopass(ctx, version, url)
-}
+	debug.Log("GPG signature OK!")
 
-func extract(archive, dest string) error {
-	debug.Log("Reading from %s", archive)
-	fh, err := os.Open(archive)
-	if err != nil {
-		return err
+	ext := "tar.gz"
+	if runtime.GOOS == "windows" {
+		ext = "zip"
 	}
-	defer func() {
-		_ = fh.Close()
-	}()
 
-	dfh, err := os.OpenFile(dest, os.O_WRONLY|os.O_CREATE|os.O_EXCL, 0755)
+	suffix := fmt.Sprintf("%s-%s.%s", runtime.GOOS, runtime.GOARCH, ext)
+	debug.Log("downloading tarball %q ...", suffix)
+	dlFilename, buf, err := downloadAsset(ctx, rel.Assets, suffix)
 	if err != nil {
-		return errors.Wrapf(err, "Failed to open file: %s", dest)
+		return err
 	}
-	defer func() {
-		_ = dfh.Close()
-	}()
 
-	gzr, err := gzip.NewReader(fh)
+	debug.Log("finding hashsum entry for %q", dlFilename)
+	wantHash, err := findHashForFile(sha256sums, dlFilename)
 	if err != nil {
 		return err
 	}
-	tarReader := tar.NewReader(gzr)
 
-	for {
-		header, err := tarReader.Next()
-		if err == io.EOF {
-			break
-		}
-		if err != nil {
-			return errors.Wrapf(err, "Failed to read from tar file")
-		}
-		name := filepath.Base(header.Name)
-		if header.Typeflag == tar.TypeReg && name == "gopass" {
-			_, err := io.Copy(dfh, tarReader)
-			return errors.Wrapf(err, "Failed to read gopass from tar file")
-		}
+	debug.Log("calculating hashsum of downloaded archive ...")
+	gotHash := sha256.Sum256(buf)
+	if !bytes.Equal(wantHash, gotHash[:]) {
+		return fmt.Errorf("SHA256 hash mismatch, want %02x, got %02x", wantHash, gotHash)
 	}
-	return errors.Errorf("file not found in archive")
-}
-
-func tryDownload(ctx context.Context, dest, url string) error {
-	bo := backoff.NewExponentialBackOff()
-	bo.MaxElapsedTime = 5 * time.Minute
-
-	return backoff.Retry(func() error {
-		select {
-		case <-ctx.Done():
-			return backoff.Permanent(errors.New("user aborted"))
-		default:
-		}
-		return download(ctx, dest, url)
-	}, bo)
-}
+	debug.Log("hashsums match!")
 
-func download(ctx context.Context, dest, url string) error {
-	fh, err := os.OpenFile(dest, os.O_WRONLY|os.O_CREATE|os.O_EXCL, 0755)
-	if err != nil {
-		return err
-	}
-	resp, err := http.Get(url)
-	if err != nil {
+	debug.Log("extracting binary from tarball ...")
+	if err := extractFile(buf, dlFilename, dest); err != nil {
 		return err
```

Comment:
So, currently the Windows updates will fail here since `extractFile` doesn't support zip files.

Reply:
I guess we should bail out early.